### PR TITLE
feat(contracts): Implement standalone freeze voting for multisig Safes

### PR DESCRIPTION
**Pull request by adamgall**
Depends on https://github.com/decentdao/decent-contracts/pull/422

Fixes [ENG-1269](https://linear.app/decent-labs/issue/ENG-1269/implement-standalone-freeze-voting-for-multisig-safes)

## Motivation

Enable token holders to freeze a multisig Safe through voting without requiring a parent/child DAO structure. This brings community oversight capabilities to standalone multisig Safes, allowing token holders to halt operations if malicious activity is detected.

## Change Summary

- Create new FreezeVotingStandaloneV1 contract for token-based freeze voting
- Implement permanent freeze model with explicit unfreeze voting
- Add security features to prevent pre-freeze transaction execution
- Create comprehensive test suite with 40+ test cases
- Move VotingConfig struct to IVotingTypes for better modularity

# Project Plan

## Current State Analysis

Multisig Safes currently lack a mechanism for token holders to intervene in case of malicious activity. The existing freeze voting contracts (FreezeVotingMultisigV1/AzoriusV1) are designed for parent/child DAO relationships and cannot be used by standalone Safes.

## Goals and Success Criteria

- ✅ Token holders can vote to freeze a multisig Safe when threshold is reached
- ✅ Frozen Safes cannot execute any transactions
- ✅ Token holders can vote to unfreeze with a separate threshold
- ✅ All pre-freeze timelocked transactions are permanently invalidated
- ✅ Support multiple voting configurations (ERC20, ERC721, etc.)
- ✅ Clean architecture without circular dependencies
- ✅ Comprehensive test coverage

## Implementation Steps

1. ✅ **Core Implementation**

   - Created FreezeVotingStandaloneV1 contract with EIP-7201 namespaced storage
   - Implemented IFreezable interface (isFrozen, lastKnownFreezeTime)
   - Added freeze/unfreeze voting mechanisms with separate thresholds
   - Integrated VotingConfig (VotingWeight + VoteTracker) architecture

2. ✅ **Security Features**

   - Added lastKnownFreezeTime() to track permanent freeze timestamp
   - Updated FreezeGuardMultisigV1 to block pre-freeze transactions
   - Prevented new timelocks during freeze
   - Created TimelockedBeforeFreeze error for clarity

3. ✅ **Interface Improvements**

   - Created IFreezable minimal interface for guards
   - Moved VotingConfig struct from IStrategyV1 to IVotingTypes
   - Updated all contracts to use IVotingTypes.VotingConfig

4. ✅ **Testing**
   - Created comprehensive test suite with 40+ test cases
   - Added tests for security invariants
   - Verified all edge cases and attack scenarios

## Benefits

- **Enhanced Security**: Token holders can intervene if multisig signers act maliciously
- **Permanent Invalidation**: All pre-freeze timelocked transactions cannot be executed
- **Flexible Voting**: Supports multiple token types through VotingConfig architecture
- **Clean Design**: No circular dependencies or complex inheritance hierarchies
- **Standalone Operation**: Works without parent/child DAO structure

## Progress Tracking

- [x] Create IFreezable interface
- [x] Update FreezeGuardMultisigV1 to use IFreezable
- [x] Implement unified freeze model (permanent freezes)
- [x] Add security features to guard
- [x] Create IFreezeVotingStandaloneV1 interface
- [x] Implement FreezeVotingStandaloneV1 contract
- [x] Write comprehensive test suite
- [x] Move VotingConfig to IVotingTypes

## Notes and Decisions

- **Permanent Freeze Model**: Freezes don't auto-expire, preventing attackers from waiting them out
- **No Bypass Mechanism**: Removed all bypass functionality for cleaner security model
- **Direct Implementation**: FreezeVotingStandaloneV1 doesn't inherit from FreezeVotingBase due to conceptual differences
- **Interface Segregation**: IFreezable reduces coupling between guards and voting contracts
- **Security Invariant**: Any transaction timelocked before a freeze is permanently invalid

### CODE REVIEW

File: contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+
+        $.unfreezeProposalVoteCount += totalWeight;
+
+        emit UnfreezeVoteCast(voter, totalWeight);
+    }
+
+    /**
+     * @inheritdoc IFreezeVotingStandaloneV1
+     */
+    function unfreeze() public virtual override(IFreezeVotingStandaloneV1) {
```
Comment (tbwebb22) [inline contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol]: I don't think we need to declare the `(IFreezeVotingStandaloneV1)` here

File: contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+        uint256 totalWeight = _getVotes(voter, votingConfigsToUse_);
+
+        // Record the vote and handle freeze activation if threshold is reached
+        _recordFreezeVote(voter, totalWeight);
+    }
+
+    /**
+     * @inheritdoc IFreezeVotingStandaloneV1
+     */
+    function castUnfreezeVote(
```
Comment (tbwebb22) [inline contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol:293]: Do we want this to automatically call `unfreeze` if the votes threshold is reached?

**Review by tbwebb22 — CHANGES_REQUESTED** [review]

File: contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol (hunk 1 above)
Comment (adamgall) [reply contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol]: good point, removed

File: contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol (hunk 2 above)
Comment (adamgall) [reply contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol:293]: yeah that's a good callout. i'll make it so that unfreezing happens automatically after unfreeze threshold is reached, to match the pattern when casting freeze votes.

Comment (adamgall) [reply contracts/deployables/freeze-voting/FreezeVotingStandaloneV1.sol:293]: ok, this is implemented and pushed up

**Review by tbwebb22 — APPROVED** [review]
